artichoke-backend/src/extn/core/time/trampoline.rs

```diff
@@ -471,11 +475,49 @@ pub fn subsec(interp: &mut Artichoke, time: Value) -> Result<Value, Error> {
 }
 
 // Time format
+fn strftime_with_encoding(
+    interp: &mut Artichoke,
+    mut time: Value,
+    format: &[u8],
+    encoding: Encoding,
+) -> Result<Value, Error> {
+    let time = unsafe { Time::unbox_from_value(&mut time, interp)? };
+
+    let bytes: Vec<u8> = time.strftime(format).map_err(|e| {
+        // InvalidFormatString is the only true ArgumentError, where as the
+        // rest that can be thrown from strftime are runtime failures.
+        //
+        // ```console
+        // [3.1.2]> Time.now.strftime("%")
+        // (irb):1:in `strftime': invalid format: % (ArgumentError)
+        //      from (irb):1:in `<main>'
+        //	    from /home/ben/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/irb-1.4.1/exe/irb:11:in `<top (required)>'
+        //	    from /home/ben/.rbenv/versions/3.1.2/bin/irb:25:in `load'
+        //	    from /home/ben/.rbenv/versions/3.1.2/bin/irb:25:in `<main>'
+        // ```
+        //
+        // Note: The errors which are re-thrown as RuntimeError include (but is
+        // not limited to: `InvalidTime`, `FormattedStringTooLarge`,
+        // `WriteZero`, `FmtError(Error)`, `OutOfMemory(TryReserveError)`
+        match e {
+            InvalidFormatString => {
+                let mut message = br#"invalid format: "#.to_vec();
+                message.extend_from_slice(format);
+                Error::from(ArgumentError::from(message))
+            }
+            _ => Error::from(RuntimeError::with_message("Unexpected failure")),
+        }
+    })?;
+
+    let result = String::with_bytes_and_encoding(bytes, encoding);
```

Comment:
For `[as]ctime` and `to_s|inspect`, it's always assumed it'll be valid utf-8.

For `Time#strftime` this might be a little bit more interesting. It's definitely making the assumption that it can enforce the coding of the provided format string. I'm mostly certain this should be a fair assumption. e.g. I assume it's impossible to create a RString with Utf8 encoding with invalid Utf8 chars right? And also, it's assuming that the `strftime-ruby` crate will never provide non ASCII chars back in place of the placeholders (This would do some weird things if the passed in format string was ASCII for example).

Do you think there should be any additional safe guards?

Reply:
It is true that `strftime-ruby` will only write ASCII bytes to for the various format specifiers.

> e.g. I assume it's impossible to create a RString with Utf8 encoding with invalid Utf8 chars right?

This is not true. `RString`s can be constructed with arbitrary bytes and any encoding. The encoding is only conventional.

It looks like when the format string is valid per its encoding, the format string's encoding is used:

```
[3.1.2] > x = "xyz %c".b
=> "xyz %c"
[3.1.2] > x.encoding
=> #<Encoding:ASCII-8BIT>
[3.1.2] > Time.now.strftime x
=> "xyz Wed Sep 21 08:57:52 2022"
[3.1.2] > Time.now.strftime(x).encoding
=> #<Encoding:ASCII-8BIT>
[3.1.2] > z = "xyz \xFF %c".b
=> "xyz \xFF %c"
[3.1.2] > z.encoding
=> #<Encoding:ASCII-8BIT>
[3.1.2] > Time.now.strftime z
=> "xyz \xFF Wed Sep 21 09:00:36 2022"
[3.1.2] > Time.now.strftime(z).encoding
=> #<Encoding:ASCII-8BIT>
```

Additionally, when the format string is not valid per its encoding, invalid bytes _and_ the encoding are preserved:

```
[3.1.2] > y = "abc \xFF %c"
=> "abc \xFF %c"
[3.1.2] > y.encoding
=> #<Encoding:UTF-8>
[3.1.2] > y.valid_encoding?
=> false
[3.1.2] > Time.now.strftime y
=> "abc \xFF Wed Sep 21 09:01:43 2022"
[3.1.2] > Time.now.strftime(y).encoding
=> #<Encoding:UTF-8>
```

Reply:
It'd be great for us to add some functional tests for the above ^. Do we have a `time_test.rb` yet?

Reply:
Ahhh yes, I see what you mean here. We could add a `time_test.rb`, however it appears there still isn't support for encodings on String yet, which makes it kind of hard to write tests for :-(.

```
artichoke 0.1.0-pre.0 (2022-09-14 revision 6305) [x86_64-unknown-linux-gnu]
[rustc 1.63.0 (4b91a6ea7 2022-08-08) on x86_64-unknown-linux-gnu]
>>> "a".b.encoding
=> #<Encoding:UTF-8>
>>> "a".force_encoding(Encoding::ASCII).encoding
=> #<Encoding:UTF-8>
```

In part due to this: https://github.com/artichoke/artichoke/blob/trunk/artichoke-backend/src/extn/core/string/string.rb#L606-L610

These might be some quick wins - I'll raise some issues for it.

Reply:
At least we can do UTF-8 and binary (`"abc".b`)